src/imitation/algorithms/dagger.py

```diff
@@ -391,30 +450,182 @@ def save_trainer(self) -> Tuple[str, str]:
+                rounded up to finish DAgger training rounds, and the
+                environment timesteps are executed in multiples of
+                `self.venv.num_envs`.)
+            rollout_round_min_episodes: The number of episodes the must be completed
+                completed before a dataset aggregation step ends.
+            rollout_round_min_timesteps: The number of environment timesteps that must
+                be completed before a dataset aggregation step ends.
+            bc_train_kwargs: Keyword arguments to pass to `BC.train` during the
+                BC update step. If this is None, then `self._default_bc_train_kwargs()`
+                is used.
+        """
+        total_timestep_count = 0
+        round_num = 0
+
+        while total_timestep_count < total_timesteps:
+            collector = self.get_trajectory_collector()
+            round_episode_count = 0
+            round_timestep_count = 0
+
+            obs = collector.reset()
+            ep_rewards = np.zeros([self.venv.num_envs])
+            # TODO(shwang): This while loop end condition causes rollout collection to
+            #   suffer from the same problem that
+            #   `imitation.data.rollout.generate_trajectories` previously had -- a
+            #   bias towards shorter episodes.
+            #   We could probably solve this problem simply by calling
+            #   `generate_trajectories(self.expert_policy, venv=collector)` now that
+            #   we are ported from using `Env` to `VecEnv`?
+            while (
+                round_episode_count < rollout_round_min_episodes
+                and round_timestep_count < rollout_round_min_timesteps
+            ):
+                acts, _ = self.expert_policy.predict(obs, deterministic=True)
+                obs, rews, dones, _ = collector.step(acts)
+                total_timestep_count += self.venv.num_envs
+                round_timestep_count += self.venv.num_envs
+                ep_rewards += rews
+
+                for i, done in enumerate(dones):
+                    if done:
+                        logger.record_mean("dagger/mean_episode_reward", ep_rewards[i])
+                        ep_rewards[i] = 0
+                round_episode_count += 1
+
+            # Flush partial trajectories so that any timesteps before episode
+            # completion are also available in the dataset.
+            # Without this call, self.extend_and_update() currently
+            # fails if the "round" also didn't finish any episodes (not enough
+            # timesteps).
+            collector.flush_trajectories()
+
+            logger.record("dagger/total_timesteps", total_timestep_count)
+            logger.record("dagger/round_num", round_num)
+            logger.record("dagger/round_episode_count", round_episode_count)
+            logger.record("dagger/round_timestep_count", round_timestep_count)
+
+            # TODO(shwang): It looks like BC might start looping Tensorboard
```

Comment:
Yes, BC loops around Tensorboard x-axis right now

![image](https://user-images.githubusercontent.com/1750835/130137203-081893e5-b401-4996-ab22-362456920df9.png)

Reply:
Could we leave this as an issue, or as a TODO comment for now? (It will require some refactoring and internal step counting for BC). I expect that tuning will be straightforward for the paper and I only need the final Sacred stats to generate the table.